src/magma/vk/physicalDevice/PhysicalDevice.cpp

```diff
@@ -0,0 +1,101 @@
+#include "magma/vk/physicalDevice/PhysicalDevice.h"
+
+#include <vector>
+#include <stdexcept>
+#include <set>
+#include <string>
+#include <sstream>
+
+#include "magma/vk/vulkan_common.h"
+
+bool QueueFamilyIndices::isComplete() {
+    return graphicsFamily.has_value() && computeFamily.has_value() && presentFamily.has_value();
+}
+
+static VkBool32 hasSurfaceSupportKHR(
+        vk::PhysicalDevice device,
+        uint32_t queueFamilyIndex,
+        vk::SurfaceKHR surface
+) {
+    return device.getSurfaceSupportKHR(queueFamilyIndex, surface).value;
+}
+
+static std::string to_string(const vk::QueueFamilyProperties& queueFamily, int i, vk::PhysicalDevice device, vk::SurfaceKHR surface) {
+    std::stringstream ss;
+    ss << i << " : " << to_string(queueFamily.queueFlags);
+    if (hasSurfaceSupportKHR(device, i, surface)) {
+        ss << " - Surface Support";
+    }
+    ss << " - " << queueFamily.queueCount;
+
+    return ss.str();
+}
+
+QueueFamilyIndices PhysicalDevice::findQueueFamilies(
+        vk::PhysicalDevice device, vk::SurfaceKHR surface
+) {
```

Comment:
May move name function and parameters on one line?

Reply:
done